### PR TITLE
Add support for HiHome radiator thermostat

### LINKED ISSUE #3922: HiHome WZB-TRVL (Smart Zigbee Radiator Thermostat) (Tuya)
## Device

- Product name: HiHome WZB-TRVL (Smart Zigbee Radiator Thermostat)
- Manufacturer: _TYST11_zuhszj9s
- Model identifier: uhszj9s
- Device type:
  - Thermostat

## Screenshots

hihome
(Note: this has been taken using this commit: https://github.com/KPWhiver/deconz-rest-plugin/commit/287a874cb026ab8c4d67e1a71141a07bb5deadde, without this commit it gets the name On/Off Light, but all the node/cluster info is the same.)

## Further information
This device is made by Tuya.
In the zigbee2mqtt project this device is implemented as a whitelabel of the Saswell-SEA802, which is also discussed at the end of this issue: https://github.com/dresden-elektronik/deconz-rest-plugin/issues/3109, and support for which has been added already.
Manufacturer: _TYST11_c88teujp
Model id: 88teujp
As such I tried to implement the HiHome thermostat myself in this commit: https://github.com/KPWhiver/deconz-rest-plugin/commit/287a874cb026ab8c4d67e1a71141a07bb5deadde but although this makes the device recognized as a thermostat in de Phoscon app and in Home Assistant, I can not control the device in Home Assistant nor do I see any temperature readings or other data.

### PR DESCRIPTION
**Pull request by KPWhiver**
This PR adds support for the HiHome (closes #3922). It also includes some fixes for the Saswell SEA801 radiator thermostat.

Manufacturer Name | Model Id | Manufacturer | Product
---------------------------|-------------|--------------------|-----------------
_TYST11_zuhszj9s | uhszj9s | HiHome |  WZB-TRVL Thermostat
_TYST11_KGbxAXL2 | GbxAXL2 | Saswell | SEA801-ZIGBEE TRV

### CODE REVIEW

**Comment by Smanar** [comment]
- Enable the defaut response for all tuya stuff (I have forget it, bad idea, provoke random bugs, realy important)
- Add support for 2 devices
- Add primary support for Tuya cluster time.
- Corrective for some Tuya schedule command.

File: tuya.cpp [hunk 1]
```diff
@@ -228,31 +227,81 @@ void DeRestPluginPrivate::handleTuyaClusterIndication(const deCONZ::ApsDataIndic
...
+                case 0x007C: // Monday
+                case 0x007D: // Thuesday
+                case 0x007E: // Wednesday
+                case 0x007F: // Thursday
+                case 0x0080: // Friday
+                case 0x0081: // Saturday
+                {
+                    int t[8] = {1,64,32,46,8,4,2};
+                    part = 1;
+                    listday << t[dp-123];
```
Comment (manup) [inline tuya.cpp:271]: This is dangerous in terms of maintenance if the code is extended later, the `dp - 123` index isn't bound checked.
Albeit due the switch case values a bound check is there, I'd prefer a more specific version. It took me a few minutes to realize that there is already an implicit array bound check and it's not a SEGV canditate.

```
const std::array<int, 7> t = {1,64,32,46,8,4,2};

if (dp < 0x007B || (dp - 0x007B) >= t.size())
{
    DPG_Printf(DBG_INFO, "Tuya unsupported daily schedule dp value: 0x%04X\n", dp);
    return; // bail out early
}
```

Note I replaced 123 by the equal hex value 0x007B to make it more clear where this lower bound offset (Sunday) comes from.

File: tuya.cpp [hunk 2]
```diff
@@ -228,31 +227,81 @@ void DeRestPluginPrivate::handleTuyaClusterIndication(const deCONZ::ApsDataIndic
...
-                    stream >> hour;
-                    stream >> minut;
-                    stream >> heatSetpoint;
+                    if (dp >= 0x007B && dp <= 0x0081)
+                    {
+                        stream >> minut16;
+                        stream >> heatSetpoint16;
+                        hour = (quint8) ((minut16 / 60) & 0xff);
+                        minut = (quint8) ((minut16 - 60 * hour) & 0xff);
+                        heatSetpoint = (quint8) ((heatSetpoint16 / 10) & 0xff);
```
Comment (manup) [inline tuya.cpp:297]: Always use C++11 casts instead old style C casts as they are nasty in terms of type safety and easily overlooked.

```cpp
hour = static_cast<quint8>((minut16 / 60) & 0xff);
minut = static_cast<quint8>((minut16 - 60 * hour) & 0xff);
heatSetpoint = static_cast<quint8>((heatSetpoint16 / 10) & 0xff);

```

File: tuya.cpp [hunk 3]
```diff
@@ -820,6 +873,51 @@ void DeRestPluginPrivate::handleTuyaClusterIndication(const deCONZ::ApsDataIndic
...
+            data.append((qint8)((time_now >> 24) & 0xff));
+            data.append((qint8)((time_now >> 16) & 0xff));
+            data.append((qint8)((time_now >> 8) & 0xff));
+            data.append((qint8)(time_now & 0xff));
+            // Ad local time
+            data.append((qint8)((time_local_time >> 24) & 0xff));
+            data.append((qint8)((time_local_time >> 16) & 0xff));
+            data.append((qint8)((time_local_time >> 8) & 0xff));
+            data.append((qint8)(time_local_time & 0xff));
+
```
Comment (manup) [inline tuya.cpp:915]: * Too much bit magic, use *QDataStream* instead;
* `QByteArray data` over shadows `quint32 data` from above;
* Always use camelCase instead snake_case when naming things, the whole code base is camel case.

```
QByteArray payload;
QDataStream stream2(&payload, QIODevice::WriteOnly);
stream2.setByteOrder(QDataStream::LittleEndian);

stream << timeNow;
stream << timeLocalTime;

```

File: tuya.cpp [hunk 4]
```diff
@@ -955,3 +1057,60 @@ bool DeRestPluginPrivate::SendTuyaRequest(TaskItem &taskRef, TaskType taskType ,
 
     return true;
 }
+
+bool DeRestPluginPrivate::SendTuyaCommand( const deCONZ::ApsDataIndication &ind, qint8 command, QByteArray data )
```
Comment (manup) [inline tuya.cpp:1061]: Don't add spaces before and after braces. &rarr; Consistency with the the rest of the code base.
Same goes for all other statements like if, for, ...

File: tuya.cpp [hunk 5]
```diff
@@ -955,3 +1057,60 @@ bool DeRestPluginPrivate::SendTuyaRequest(TaskItem &taskRef, TaskType taskType ,
...
+    TaskItem task;
+    
+    //Tuya task
+    task.taskType = TaskTuyaRequest;
+    
+    task.req.dstAddress() = ind.srcAddress();
+    task.req.setDstAddressMode(deCONZ::ApsExtAddress);
+    task.req.setDstEndpoint(ind.srcEndpoint() );
+    task.req.setSrcEndpoint( endpoint() );
+    task.req.setDstEndpoint(ind.srcEndpoint() );
```
Comment (manup) [inline tuya.cpp:1074]: Spaces...

File: tuya.cpp [hunk 6]
```diff
@@ -955,3 +1057,60 @@ bool DeRestPluginPrivate::SendTuyaRequest(TaskItem &taskRef, TaskType taskType ,
...
+    // payload
+    QDataStream stream(&task.zclFrame.payload(), QIODevice::WriteOnly);
+    stream.setByteOrder(QDataStream::LittleEndian);
+    
+    // Data
+    stream << (qint16) data.length(); // length always 8
+    for (int i = 0; i < data.length(); i++)
+    {
+        stream << (quint8) data[i];
+    }
```
Comment (manup) [inline tuya.cpp:1094]: Use C++11 `static_cast<Type>(Value)`

File: tuya.cpp [hunk 7]
```diff
@@ -955,3 +1057,60 @@ bool DeRestPluginPrivate::SendTuyaRequest(TaskItem &taskRef, TaskType taskType ,
...
+        task.zclFrame.writeToStream(stream);
+    }
+
+    if (addTask(task))
+    {
+        taskToLocalData(task);
+    }
+    else
+    {
+        DBG_Printf(DBG_INFO, "Tuya debug 7");
```
Comment (manup) [inline tuya.cpp:1109]: Don't print generic debug messages, it doesn't say anything meaningful.

```
DBG_Printf(DBG_INFO, "Failed to send Tuya command 0x%02X data: %s\n", command, qPrintable(data.toHex()));
```

**Review by manup — CHANGES_REQUESTED** [review]
Added some comments @Smanar can you please check them.

**Comment by Smanar** [comment]
Still on it, pls do not merge yet.

**Comment by Smanar** [comment]
@KPWhiver I have updated the code, if you have some time, to test it again ?

**Comment by manup** [comment]
>It also includes some fixes for the Saswell SEA801 radiator thermostat.

Please also list the "some fixes" in the PR description so readers know what the PR fixes :)

**Comment by Smanar** [comment]
From my memory it's from here > https://github.com/dresden-elektronik/deconz-rest-plugin/issues/3922
Without the defaut response, some command was missing from the device, random bug, random commands.

**Comment by KPWhiver** [comment]
Yes, which brings me to another problem, the HiHome thermostat is a whitelabel of the SEA802 thermostat, which means it should be implemented identically. Unfortunately it isn't which makes me strongly suspect that SEA802 support in deconz is actually not there yet. But I don't think it's wise to also fix that here, since I don't have a SEA802 to test it with.

I'll test the fixes.

**Comment by Smanar** [comment]
@KPWhiver it s the model "_TYST11_KGbxAXL2" ?
If I remember this model need to mimic the  "_TYST11_zuhszj9s" model, so easy to see if something is missing in code. I have check it have less feature enabled (no window closed check for exemple), but thoses 2 devices are "friends"

File: database.cpp [hunk 8]
```diff
@@ -3560,18 +3564,30 @@ static int sqliteLoadAllSensorsCallback(void *user, int ncols, char **colval , c
...
                    (sensor.manufacturer() == QLatin1String("_TZE200_aoclfnxz")) )   // Tuya
                 {
-                    sensor.addItem(DataTypeString, RConfigPreset);
                     sensor.addItem(DataTypeBool, RConfigLocked)->setValue(false);
+                }
+
+                if (sensor.modelId() == QLatin1String("kud7u2l") || // Tuya
+                    sensor.modelId() == QLatin1String("eaxp72v") || // Tuya
+                    sensor.modelId() == QLatin1String("fvq6avy") || // Tuya
+                    sensor.modelId() == QLatin1String("88teujp") || // Tuya
```
[the remark this answers is not in the document]
Comment (Smanar) [reply database.cpp]: Wiki page started > https://github.com/dresden-elektronik/deconz-rest-plugin/wiki/Tuya-devices-List

**Comment by KPWhiver** [comment]
No it's the _TYST11_c88teujp model, it should be identical to the HiHome thermostat. 

I tried checking out this pull request and running it but when I do deconz won't startup properly, this is what I get in the logs:
```
Jan 26 22:24:08 server conmon[26980]: New 'server:0 (root)' desktop at :0 on machine server
Jan 26 22:24:08 server conmon[26980]:
Jan 26 22:24:08 server conmon[26980]: Starting applications specified in /etc/X11/Xvnc-session
Jan 26 22:24:08 server conmon[26980]: Log file is /root/.vnc/server:0.log
Jan 26 22:24:08 server conmon[26980]:
Jan 26 22:24:08 server conmon[26980]: Use xtigervncviewer -SecurityTypes VncAuth,TLSVnc -passwd /root/.vnc/passwd server:0 to connect to the VNC server.
Jan 26 22:24:08 server conmon[26980]:
Jan 26 22:24:11 server conmon[26980]: QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
Jan 26 22:24:11 server conmon[26980]: libpng warning: iCCP: known incorrect sRGB profile
Jan 26 22:24:11 server conmon[26980]: libpng warning: iCCP: known incorrect sRGB profile
Jan 26 22:24:11 server conmon[26980]: 22:24:11:993 HTTP Server listen on address 0.0.0.0, port: 8080, root: /usr/share/deCONZ/webapp/
Jan 26 22:24:12 server conmon[26980]: 22:24:12:014 CTRL. 3.27.2
Jan 26 22:24:12 server conmon[26980]: 22:24:12:049 ZCLDB init file /root/.local/share/dresden-elektronik/deCONZ/zcldb.txt
Jan 26 22:24:12 server conmon[26980]: 22:24:12:115 found node plugin: libde_signal_plugin.so - Signal Monitor Plugin
Jan 26 22:24:12 server conmon[26980]: 22:24:12:119 found node plugin: libstd_otau_plugin.so - STD OTAU Plugin
Jan 26 22:24:13 server conmon[26980]: 22:24:13:355 COM check bootloader
Jan 26 22:24:13 server conmon[26980]: QObject::connect: Cannot queue arguments of type 'QTextBlock'
Jan 26 22:24:13 server conmon[26980]: (Make sure 'QTextBlock' is registered using qRegisterMetaType().)
Jan 26 22:24:13 server conmon[26980]: QObject::connect: Cannot queue arguments of type 'QTextCursor'
Jan 26 22:24:13 server conmon[26980]: (Make sure 'QTextCursor' is registered using qRegisterMetaType().)
Jan 26 22:24:13 server conmon[26980]: 22:24:13:367 COM detected application
Jan 26 22:24:13 server conmon[26980]: 22:24:13:398 Device firmware version 0x26660700 ConBee II
Jan 26 22:24:13 server conmon[26980]: 22:24:13:402 unlocked max nodes: 200
Jan 26 22:24:13 server conmon[26980]: 22:24:13:473 Device protocol version: 0x010C
Jan 26 22:24:13 server conmon[26980]: 22:24:13:481 new node - ext: 0x00212effff05412d, nwk: 0x0000
Jan 26 22:24:13 server conmon[26980]: 22:24:13:582 Current channel 15
Jan 26 22:24:13 server conmon[26980]: 22:24:13:598 CTRL ANT_CTRL 0x03
Jan 26 22:24:13 server conmon[26980]: 22:24:13:623 Device protocol version: 0x010C
Jan 26 22:24:13 server conmon[26980]: 22:24:13:671 Current channel 15
Jan 26 22:24:13 server conmon[26980]: 22:24:13:687 CTRL ANT_CTRL 0x03
Jan 26 22:25:11 server conmon[26980]: 22:25:11:808 Current channel 15
Jan 26 22:25:11 server conmon[26980]: 22:25:11:816 Device TTL 6671 s flags: 0x7
Jan 26 22:26:11 server conmon[26980]: 22:26:11:808 Current channel 15
Jan 26 22:26:11 server conmon[26980]: 22:26:11:816 Device TTL 6611 s flags: 0x7
Jan 26 22:27:11 server conmon[26980]: 22:27:11:807 Current channel 15
Jan 26 22:27:11 server conmon[26980]: 22:27:11:815 Device TTL 6551 s flags: 0x7
Jan 26 22:28:11 server conmon[26980]: 22:28:11:809 Current channel 15
Jan 26 22:28:11 server conmon[26980]: 22:28:11:817 Device TTL 6490 s flags: 0x7
Jan 26 22:29:11 server conmon[26980]: 22:29:11:808 Current channel 15
Jan 26 22:29:11 server conmon[26980]: 22:29:11:816 Device TTL 6430 s flags: 0x7
Jan 26 22:30:11 server conmon[26980]: 22:30:11:808 Current channel 15
Jan 26 22:30:11 server conmon[26980]: 22:30:11:816 Device TTL 6370 s flags: 0x7
Jan 26 22:31:11 server conmon[26980]: 22:31:11:808 Current channel 15
Jan 26 22:31:11 server conmon[26980]: 22:31:11:816 Device TTL 6310 s flags: 0x7
Jan 26 22:32:11 server conmon[26980]: 22:32:11:807 Current channel 15
Jan 26 22:32:11 server conmon[26980]: 22:32:11:815 Device TTL 6250 s flags: 0x7
Jan 26 22:33:11 server conmon[26980]: 22:33:11:809 Current channel 15
Jan 26 22:33:11 server conmon[26980]: 22:33:11:817 Device TTL 6189 s flags: 0x7
Jan 26 22:34:11 server conmon[26980]: 22:34:11:808 Current channel 15
Jan 26 22:34:11 server conmon[26980]: 22:34:11:816 Device TTL 6129 s flags: 0x7
Jan 26 22:35:11 server conmon[26980]: 22:35:11:809 Current channel 15
Jan 26 22:35:11 server conmon[26980]: 22:35:11:817 Device TTL 6069 s flags: 0x7
Jan 26 22:36:11 server conmon[26980]: 22:36:11:807 Current channel 15
Jan 26 22:36:11 server conmon[26980]: 22:36:11:815 Device TTL 6009 s flags: 0x7
Jan 26 22:37:11 server conmon[26980]: 22:37:11:809 Current channel 15
Jan 26 22:37:11 server conmon[26980]: 22:37:11:817 Device TTL 5949 s flags: 0x7
Jan 26 22:38:11 server conmon[26980]: 22:38:11:809 Current channel 15
Jan 26 22:38:11 server conmon[26980]: 22:38:11:817 Device TTL 5889 s flags: 0x7
Jan 26 22:39:11 server conmon[26980]: 22:39:11:809 Current channel 15
Jan 26 22:39:11 server conmon[26980]: 22:39:11:817 Device TTL 5828 s flags: 0x7
Jan 26 22:40:11 server conmon[26980]: 22:40:11:808 Current channel 15
Jan 26 22:40:11 server conmon[26980]: 22:40:11:816 Device TTL 5768 s flags: 0x7
etc ...
```

**Comment by Smanar** [comment]
"_TYST11_c88teujp" no problem for this one, better integration it is exactly at same place than the one finishing by uhszj9s

I have already have this problem, in epoch code, the compiler don't give error message, but the code is not working.

**Comment by Smanar** [comment]
@KPWhiver corrected, the code compile and work now, on the previous one, the REST API was missing if you look at deconz menu.

**Comment by KPWhiver** [comment]
With the fixes, deconz now runs, and I can send commands to the thermostats. However, they do not send any temperatures back, only the valve state:
```
Jan 28 08:01:27 server conmon[4063]: 08:01:27:960 Tuya debug 4 : Address 0x60A423FFFE8E2CE7 Payload 00e96501000101
Jan 28 08:00:28 server conmon[4063]: 08:00:28:015 Tuya debug 5 : Status: 0 Transid: 233 Dp: 357 (0x01,0x65) Fn: 0 Data 1
Jan 28 08:00:28 server conmon[4063]: 08:00:28:015 Tuya debug 4 : Address 0x60A423FFFE8E2CE7 Payload 00e96501000101
Jan 28 07:59:27 server conmon[4063]: 07:59:27:966 Tuya debug 5 : Status: 0 Transid: 233 Dp: 357 (0x01,0x65) Fn: 0 Data 1
Jan 28 07:59:27 server conmon[4063]: 07:59:27:963 Tuya debug 4 : Address 0x60A423FFFE8E2CE7 Payload 00e96501000101
Jan 28 07:59:18 server conmon[4063]: 07:59:18:245 Tuya debug 5 : Status: 0 Transid: 233 Dp: 357 (0x01,0x65) Fn: 0 Data 1
Jan 28 07:59:18 server conmon[4063]: 07:59:18:245 Tuya debug 4 : Address 0x60A423FFFE8E2CE7 Payload 00e96501000101
Jan 28 07:58:12 server conmon[4063]: 07:58:12:956 Tuya debug 5 : Status: 0 Transid: 233 Dp: 357 (0x01,0x65) Fn: 0 Data 1
Jan 28 07:58:12 server conmon[4063]: 07:58:12:956 Tuya debug 4 : Address 0x60A423FFFE8E2CE7 Payload 00e96501000101
Jan 28 07:57:12 server conmon[4063]: 07:57:12:957 Tuya debug 5 : Status: 0 Transid: 233 Dp: 357 (0x01,0x65) Fn: 0 Data 1
Jan 28 07:57:12 server conmon[4063]: 07:57:12:956 Tuya debug 4 : Address 0x60A423FFFE8E2CE7 Payload 00e96501000101
Jan 28 07:56:12 server conmon[4063]: 07:56:12:960 Tuya debug 5 : Status: 0 Transid: 233 Dp: 357 (0x01,0x65) Fn: 0 Data 1
Jan 28 07:56:12 server conmon[4063]: 07:56:12:959 Tuya debug 4 : Address 0x60A423FFFE8E2CE7 Payload 00e96501000101
Jan 28 07:55:31 server conmon[4063]: 07:55:31:504 Tuya debug 5 : Status: 0 Transid: 233 Dp: 357 (0x01,0x65) Fn: 0 Data 1
Jan 28 07:55:31 server conmon[4063]: 07:55:31:504 Tuya debug 4 : Address 0x60A423FFFE8E2CE7 Payload 00e96501000101
Jan 28 07:55:12 server conmon[4063]: 07:55:12:303 Tuya debug 5 : Status: 0 Transid: 233 Dp: 357 (0x01,0x65) Fn: 0 Data 1
Jan 28 07:55:12 server conmon[4063]: 07:55:12:303 Tuya debug 4 : Address 0x60A423FFFE8E2CE7 Payload 00e96501000101
Jan 28 07:55:02 server conmon[4063]: 07:55:02:207 Tuya debug 5 : Status: 0 Transid: 233 Dp: 357 (0x01,0x65) Fn: 0 Data 1
```

This suggests that for some reason the default response is not send anymore.

**Comment by Smanar** [comment]
Oup, big thx for tests, all was broken, all the code was bypassed on first test.

**Comment by KPWhiver** [comment]
It has been running stable for a day now, everything seems to work

File: database.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ebaauw) [reply database.cpp]: @manup, @Smanar Why a Wiki page?  Or rather: why not a table in code, like I've done for the LIDL devices?
https://github.com/dresden-elektronik/deconz-rest-plugin/blob/f46e3321715f755a057c19992e51ce9cf6f01e35/de_web_plugin.cpp#L447-L460

[the remark this answers is not in the document]
Comment (Smanar) [reply database.cpp]: To be more accessible ?
I think the table can be moved soon with that https://github.com/dresden-elektronik/deconz-rest-plugin/pull/4301/files

**Comment by k3mpaxl** [comment]
Hi @Smanar, 
Time synchronization is broken. Time responses are being sent, but the time displayed jumps randomly. I'll try to sniff the traffic and report back to you.

**Comment by Smanar** [comment]
Ha :(
But I don't see a change with the version from the other issue ...

**Comment by k3mpaxl** [comment]
The time syncronization is broken, the byte order in the response is wrong, but the header byte order is correct!

Client request:
sniff-03
Response with this implementation.
sniff-04

It should look like this:
```
0000   08 04 60 19 a7 7c 60 19 b5 8c
```

**Comment by Smanar** [comment]
Ha yes I m seing the modification

Before 

```
        //Add the "magic vlaue"
        data.append((qint8)(UnknowHeader & 0xff));
        data.append((qint8)((UnknowHeader >> 8) & 0xff));
        
        // Add UTC time
        data.append((qint8)((time_now >> 24) & 0xff));
        data.append((qint8)((time_now >> 16) & 0xff));
        data.append((qint8)((time_now >> 8) & 0xff));
        data.append((qint8)(time_now & 0xff));
        // Ad local time
        data.append((qint8)((time_local_time >> 24) & 0xff));
        data.append((qint8)((time_local_time >> 16) & 0xff));
        data.append((qint8)((time_local_time >> 8) & 0xff));
        data.append((qint8)(time_local_time & 0xff));
```

after
```
        //Add the "magic value"
        stream2 << UnknowHeader;
         // Add UTC time
        stream2 << time_now;
        // Ad local time
        stream2 << time_local_time;
```

Have done a change, if you can test it pls ? https://github.com/KPWhiver/deconz-rest-plugin/commit/86a64137cf9d8f08564aaa78906de298168071b1

**Comment by k3mpaxl** [comment]
> 
> 
> Ha yes I m seing the modification
> 
> Before
> 
> ```
>         //Add the "magic vlaue"
>         data.append((qint8)(UnknowHeader & 0xff));
>         data.append((qint8)((UnknowHeader >> 8) & 0xff));
>         
>         // Add UTC time
>         data.append((qint8)((time_now >> 24) & 0xff));
>         data.append((qint8)((time_now >> 16) & 0xff));
>         data.append((qint8)((time_now >> 8) & 0xff));
>         data.append((qint8)(time_now & 0xff));
>         // Ad local time
>         data.append((qint8)((time_local_time >> 24) & 0xff));
>         data.append((qint8)((time_local_time >> 16) & 0xff));
>         data.append((qint8)((time_local_time >> 8) & 0xff));
>         data.append((qint8)(time_local_time & 0xff));
> ```
> 
> after
> 
> ```
>         //Add the "magic value"
>         stream2 << UnknowHeader;
>          // Add UTC time
>         stream2 << time_now;
>         // Ad local time
>         stream2 << time_local_time;
> ```
> 
> Have done a change, if you can test it pls ? [KPWhiver@86a6413](https://github.com/KPWhiver/deconz-rest-plugin/commit/86a64137cf9d8f08564aaa78906de298168071b1)

Time synchronization is now working fine! Thanks for the late change.

**Comment by Smanar** [comment]
@k3mpaxl Thx to you ^^.
For information I have no tuya devices at all, all the code was done with helpfull guys like you, I realy can't do anything without you.

File: tuya.cpp [hunk 9]
```diff
@@ -860,6 +989,10 @@ bool DeRestPluginPrivate::SendTuyaRequestThermostatSetWeeklySchedule(TaskItem &t
     {
         //To finish
     }
+    else if (Dp_identifier == 0x6D)
```
Comment (manup) [inline tuya.cpp:992]: Should the 0x6D be `DP_IDENTIFIER_THERMOSTAT_SCHEDULE_4`?

**Review by manup — CHANGES_REQUESTED** [review]
Next round, added a few comments.
In general I really like the new defines like DP_IDENTIFIER_THERMOSTAT_SCHEDULE_2
makes the code more readable :+1:

**Comment by manup** [comment]
@KPWhiver  small note, I'll add my requested changes in a further PR to get the release ready. Mostly changes are related to code style so logic will be the same.

**Pull request merged**